src/lib.rs

```diff
@@ -0,0 +1,27 @@
+use std::fs;
+use gitignore;
+
+pub fn print_gtree(entries: fs::ReadDir, gitignore: Option<gitignore::File>) -> std::io::Result<()> {
+    for entry in entries {
+        let dir = entry?;
+        let mut ignore = false;
+
+        if let Some(g) = &gitignore {
+            match g.is_excluded(&dir.path()) {
+                Ok(true) => {
+                    ignore = true;
+                },
+                Ok(false) => {
+                },
+                Err(e) => {
+                    println!("failing:{:?}", e);
+                }
+            }
+        }
```

Comment:
Instead of this explicit `match` on the status and mutating the `ignore` variable, you can use a couple useful APIs on `Result` and `Option` that you might not be used to from Elm.

First, [`Option::as_ref`](https://doc.rust-lang.org/std/option/enum.Option.html#method.as_ref) takes an owned `Option<T>` by `&self` and returns an `Option<&T>`.  [`Option::and_then`](https://doc.rust-lang.org/std/option/enum.Option.html#method.and_then) allows you to chain `Option`al values together taking a fn that takes the `Some` value as an argument, and [`Result::ok`](https://doc.rust-lang.org/std/result/enum.Result.html#method.ok) turns a `Result<T, E>` into an `Option<T>`. Finally, we have [`Option::unwrap_or`](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or) which either produces the value contained in a `Some` or the default value which is the second argument.

Using all those you should be able to write a single `if` statement that has the two `println`s currently in this function in each of the two branches. Since this method chain is likely to be a little bulky I would pull it out into a helper function too.

Reply:
The REALLY cool thing about this is that a simpler version of this helper function would then be mappable across a `Vec<gitignore::File>` using [`Iterator::any`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.any).